src/torch_models.py

```diff
@@ -364,26 +363,43 @@ def _initialize_net(self) -> None:
         self._linears.append(nn.Linear(self._hid_sizes[-1], 1))
 
     def _create_loss_fn(self) -> Callable[[Tensor, Tensor], Tensor]:
-        return nn.BCEWithLogitsLoss()
+
+        def _loss_fn(Y_hat: Tensor, Y: Tensor) -> Tensor:
+            pred = Y_hat.reshape(Y.shape)  # (num data, num classes)
+            log_probs = F.log_softmax(pred / self._temperature, dim=-1)
+            loss = F.kl_div(log_probs, Y, reduction='batchmean')
+            return loss
+
+        return _loss_fn
 
     def _create_batch_generator(self, X: Array,
                                 Y: Array) -> Iterator[Tuple[Tensor, Tensor]]:
-        # Resample negative examples on each iteration.
-        pos_concat_inputs = np.hstack([X, Y])
-        num_pos_inputs = len(pos_concat_inputs)
-        num_neg_inputs = num_pos_inputs * self._num_negative_data_per_input
-        targets = np.array([1] * num_pos_inputs + [0] * num_neg_inputs,
-                           dtype=np.float32)
-        tensor_Y = torch.from_numpy(np.array(targets, dtype=np.float32))
+        num_samples = X.shape[0]
+        num_negatives = self._num_negatives_per_input
+        # Cast to torch first.
+        tensor_X = torch.from_numpy(np.array(X, dtype=np.float32))
+        tensor_Y = torch.from_numpy(np.array(Y, dtype=np.float32))
+        # Expand tensor_Y in preparation for concat in the loop below.
+        tensor_Y = tensor_Y[:, None, :]
+        # For each of the negative outputs, we need a corresponding input.
+        tiled_X = tensor_X.unsqueeze(1).repeat(1, num_negatives + 1, 1)
+        extended_X = tiled_X.reshape([-1, tensor_X.shape[-1]])
         while True:
-            neg_X, neg_Y = self._create_negative_data(X, Y)
-            # Set up the data for the classifier.
-            neg_concat_inputs = np.hstack([neg_X, neg_Y])
-            concat_inputs = np.vstack([pos_concat_inputs, neg_concat_inputs])
-            # Convert data to tensors.
-            tensor_X = torch.from_numpy(
-                np.array(concat_inputs, dtype=np.float32))
-            yield (tensor_X, tensor_Y)
+            # Resample negative examples on each iteration.
+            neg_Y = torch.rand(size=(num_samples, num_negatives, self._y_dim),
+                               dtype=tensor_Y.dtype)
+            # Create a multiclass classification-style target vector.
+            combined_Y = torch.cat([tensor_Y, neg_Y], axis=1)  # type: ignore
+            combined_Y = combined_Y.reshape([-1, tensor_Y.shape[-1]])
```

Comment:
i don't understand what this line is doing. can you go through a concrete example of the shapes?
so tensor_X is (num_samples, state_dim)
and tensor_Y is (num_samples, action_dim)
we are creating num_neg_samples negative examples of actions, so combined_Y should be (num_samples + num_neg_samples, action_dim)
then what's the reshape doing? and why is the `neg_Y = ...` 3-dimensional? i'm clearly misunderstanding something...